python_modules/dagster/dagster/_daemon/concurrency.py

```diff
@@ -0,0 +1,46 @@
+import logging
+from typing import Iterator, Optional
+
+import pendulum
+
+from dagster._core.storage.dagster_run import (
+    FINISHED_STATUSES,
+    RunsFilter,
+)
+from dagster._core.workspace.context import IWorkspaceProcessContext
+from dagster._utils import DebugCrashFlags
+from dagster._utils.error import SerializableErrorInfo
+
+
+def execute_concurrency_slots_iteration(
+    workspace_process_context: IWorkspaceProcessContext,
+    logger: logging.Logger,
+    _debug_crash_flags: Optional[DebugCrashFlags] = None,
+) -> Iterator[Optional[SerializableErrorInfo]]:
+    instance = workspace_process_context.instance
+    timeout_seconds = instance.run_monitoring_settings.get("free_concurrency_slots_seconds")
+    if not timeout_seconds:
+        yield
+        return
+
+    run_ids = instance.event_log_storage.get_concurrency_run_ids()
+    if not run_ids:
+        yield
+        return
+
+    run_records = instance.get_run_records(filters=RunsFilter(run_ids=list(run_ids)))
```

Comment:
How big do we except this list to get on cloud?

Reply:
This is limited to the number of in-progress runs.   I think this seems reasonable, but I could add a pretty high limit on it if you have concerns.